init BJS2-65635

RamilNazmiev192 opened the pull request.

RamilNazmiev192 committed:
init BJS2-65635

EvgenSM10 commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,58 @@
+import org.telegram.telegrambots.updatesreceivers.DefaultBotSession;
+
+@Service
+@RequiredArgsConstructor
```
аннотация не нужна

RamilNazmiev192 replied:
done

RamilNazmiev192 replied:
создал TelegramProperties
пригодился

EvgenSM10 commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,58 @@
+    private MyTelegramBot bot;
+
+    @PostConstruct
+    public void init() throws TelegramApiException {
```
здесь нужно добавить валидацию botToken и botUserName

RamilNazmiev192 replied:
done

EvgenSM10 commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,58 @@
+
+    @PostConstruct
+    public void init() throws TelegramApiException {
+        TelegramBotsApi botsApi = new TelegramBotsApi(DefaultBotSession.class);
```
я думаю здесь в try-catch нужно обернуть создание бота

RamilNazmiev192 replied:
Spring поймает его при старте приложения.
ведь без Telegram бот не работает, и запускать сервис без него не имеет смысла.

EvgenSM10 commented on src/main/java/faang/school/notificationservice/service/telegram/MyTelegramBot.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public void onUpdateReceived(Update update) {
+        if (update.hasMessage()) {
+            System.out.println("Telegram Chat ID: " + update.getMessage().getChatId());
```
лучше через log

RamilNazmiev192 replied:
done

EvgenSM10 commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        SendMessage sendMessage = new SendMessage();
+        sendMessage.setChatId(String.valueOf(user.getTelegramChatId()));
```
String.valueof лишний, TelegramChatId имеет тип String

RamilNazmiev192 replied:
done

EvgenSM10 commented on src/main/java/faang/school/notificationservice/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+public class GlobalExceptionHandler {
+    @ExceptionHandler(IllegalArgumentException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleSmsIntegrationException(IllegalArgumentException e) {
```
handleIllegalArgumentException

RamilNazmiev192 replied:
done

RamilNazmiev192 committed:
refactoring праваки после кросс ревю

Soddik commented on src/main/java/faang/school/notificationservice/config/context/TelegramProperties.java:
```diff
@@ -0,0 +1,20 @@
+@Setter
+@Validated
+@ConfigurationProperties(prefix = "telegram")
+public class TelegramProperties {
```
лучше переименовать в TelegramBotProperties

RamilNazmiev192 replied:
done

Soddik commented on src/main/java/faang/school/notificationservice/handler/ErrorResponse.java:
```diff
@@ -0,0 +1,33 @@
+public class ErrorResponse {
+
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime timestamp;
```
всегда null?

RamilNazmiev192 replied:
done

Soddik commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,59 @@
+
+    private final TelegramProperties telegramProperties;
+    @VisibleForTesting
+    MyTelegramBot bot;
```
почему не private? 
можно через инжект в конструктор?

RamilNazmiev192 replied:
1 я потратил час на тестирование и это самый оптимальный вариант который  я нашел, другие способы не рабают (ну или мне не хватает знаний) , могу сделать private но тогда без тестов

Soddik commented on src/main/java/faang/school/notificationservice/service/telegram/TelegramService.java:
```diff
@@ -0,0 +1,59 @@
+
+        try {
+            bot.execute(sendMessage);
+        } catch (Exception e) {
```
перехватываются все исключения, лучше ловить конкретные, а Exception обернуть в ControllerAdvice

RamilNazmiev192 replied:
Согласен, ловить Exception в бизнес-логике не best practice, если это не обосновано.
В моём случае bot.execute(...) потенциально может бросить любые RuntimeException, включая TelegramApiException, но их иерархия не всегда очевидна, поэтому я решил перехватить Exception, логировать и обернуть в NotificationException.

При этом у нас есть @ControllerAdvice, который уже обрабатывает NotificationException централизованно — то есть, от клиента не скрывается ошибка, и стек логируется.

Если появится более конкретный тип исключения (например, TelegramApiException или IOException), — я без проблем добавлю конкретный catch.

RamilNazmiev192 committed:
refactoring
1 заполнение тело ошибок
2 добовление глобалного обработчика
3 переименовал TelegramProperties
